### ISSUE TITLE
Increase the default size of threadpool:automation's default

### ISSUE THREAD

**Issue by rkoshak** [openhab/openhab-core#1889]
I wish I had seen this issue/PR go by when it happened but it seems to be a _fait accompli_. Unfortunately, this feels like a place where the efficiency of the computer took precedence over the human usability. 

One issue we spend a huge amount of time on the forum dealing with is users running out of rules threads. It's not just people misusing Thread.sleeps. Normal and expected calls like `executeCommandLine` and all the `sendHttpXRequest` calls can take seconds or more to complete. Over and over again users run into this problem. I was really looking forward to a world where I wouldn't have to keep helping people with this same problem. And given that we don't require openHAB users to be knowledgeable about parallel programming, the fact that there is a problem and it's cause is almost completely unexpected and counter intuitive to them.

With the introduction of the NGRE, apparently it's been using one thread per rule for all this time. I suspect some early adopters like myself have taken note of that and relaxed the strict requirement to create short running rules (sometime to the extent of outsourcing stuff that would easily be handled by executeCommandLine to an external script that calls the command based on receipt of an MQTT message). That's a heavy load to lay on someone who's first exposure to programming is openHAB rules.

So my requests are two fold:

1. This needs to be announced as a breaking change. To this point those who have been using automation other than Rules DSL in OH 2.5 have known that the limited thread pool isn't an issue and may have relaxed on long running rules. The new thread pool has a high probability of breaking their rules.

2. Increase the default size of the threadpool. A default of 5 seems *way* to low. Increasing it to 10 or 20 would probably eliminate 90% of all the problems users will run into when making calls to executeCommandLine et. al. Even better would be to use a thread pool that can grow dynamically to some maximum size (much larger than 5). In that case most users would probably live within the optimum minimum, but OH itself will adjust to those users who happen to have a lot going on at the same time or calls to long running rules which may not be avoidable. Then you could have something like a max size of 50 or something crazy like that and only those users who are really abusing rules would encounter problems. 

And one more question that I haven't been able to determine on my own yet. It's still the case that only one instance of a given rule can run at a time, right? We are not back into the situation where you can have more than one copies of a rule running at the same time I hope. This could be a particularly big deal with ECMASCript 5.1 rules in the JSONDB because, based on my observations, the same global context gets reused and if two copies of the rule are running at the same time one of two things would occur:

- the second instance gets it's own copy of the context so it won't see all the timers and stuff that are stored there (like global variables in Rules DSL)
- they both get the same context in which case they will stomp all over each other's global variables.

In the second case the solution would be to implement a rule lock or locked portions of rules which is just going to exacerbate the problem caused by the small threadpool size. And it's a complexity that will be hard for many of our new users to grasp.

**Comment by kaikreuzer** [comment]
I see your point @rkoshak.

I'd also love to be able to define a large (e.g. 50 or 100) maximum pool size, which simply acts as an upper bound to avoid OOM errors and other resource leaks.
Unfortunately, we never found a way to make the thread pools behave that way. Instead, they always create new threads as long as they haven't reached their limit, instead of reusing idle threads. Only once they run into an idle timeout, they are discarded, which means that new threads will be immediately created, if new jobs come in.
I've just tested it for rules: Setting it to 100 means that there are 100 threads open all the time. And even worse: Some are discarded, while new ones are spawned all the time...
So I indeed wouldn't set it to more than 20 (with this size, I actually do not see them being discarded and recreated - they stay stable).
I'd also be open to revert https://github.com/openhab/openhab-core/pull/1857 again, if it causes us too much trouble.
The situation then is that we have one permanent thread per rule that has some item trigger - maybe not so bad after all. People with many rules will have many threads, people with few rules only a few. 

So what do you prefer? Setting threadpool to 20 or reverting #1857?

>  It's still the case that only one instance of a given rule can run at a time, right?

Absolutely right and I had to learn that this week as well, see https://github.com/openhab/openhab-core/issues/1868#issuecomment-735469672. It absolutely makes sense to keep every rule being single-threaded as everything else will cause havoc at many places...

**Comment by morph166955** [comment]
As a person who runs into this issue frequently, I would like to chime in here.  When I was on 3.0.0M2, I noticed my rules ran amazingly.  Things that always had problems running in parallel just did.  I didn't have to set anything on the threadpools at all, it just worked.  When I moved to 3.0.0M4 (or more realistically 2091 after that last bug), I noticed my rules weren't running in parallel anymore.  As a test I increased my pool to 20.  It helped, but it wasn't as perfect as it was in M2.  If I had a vote, I would go for reverting #1857 over increasing the size of the pool.

**Comment by kaikreuzer** [comment]
@morph166955 Noted, thanks.

@openhab/core-maintainers Any preferences on your side?

**Comment by rkoshak** [comment]
Not knowing what the real impact of having one thread per rule sitting there idle most of the time I'd say revert #1857 back and let there be one thread per rule. But it's not a strong preference. 20 threads ought to be enough to head off most of the problems that bring users to the forum.

I do understand how it feels icky to have so much memory devoted to threads that are not doing anything most of the time. But in this case I think the benefits to the user justify the expense. I don't have enough rules on OH 3 yet to provide any before and after experience like @morph166955 but that seems to be another pretty compelling bit of anecdote to support reverting.

**Comment by kaikreuzer** [comment]
Ok, thanks for your input. Assuming the other core maintainers agree as well, I have prepared https://github.com/openhab/openhab-core/pull/1890.

**Comment by Nadahar** [comment]
> Unfortunately, we never found a way to make the thread pools behave that way. Instead, they always create new threads as long as they haven't reached their limit, instead of reusing idle threads. Only once they run into an idle timeout, they are discarded, which means that new threads will be immediately created, if new jobs come in

@kaikreuzer I know this is old, but this is still the situation today, and I think I can explain why you can't get dynamic pools "to behave". The reason is the way OH uses the "core threads" in a thread pool. The core threads are supposed to be the bare minimum that is always running in a pool, it might just be one thread for many pools. They aren't supposed to time out. Pools are still possible to configure so that core threads time out, but that's meant for the more odd scenario. As far as I can see, OH do this "as standard".

Core threads and "other threads" follow different rules. Each time a task is submitted to the queue, if the number of running threads is below the core thread count, one new thread is spawned. This only happens until you have the defined number of core threads running. After that, threads will only spawn on demand, and idling threads not yet shut down will pick up new tasks - like you'd expect. This works fine, I've used it many times in Java. The pool adjusts dynamically and generally only run the number of threads needed. The number of core thread should just be the very minimum that should be running for the pool at any given moment, and they should normally not be set to time-out, it's better to just have a singe core thread then.

The "challenge" (and major mistake in Java's implementation of thread pools in my opinion), it the way new threads are spawned. A pool will only spawn new threads when the queue is *full*. There is no way to configure a certain "load" or something, it has to be 100% full. As a consequence, dynamic pools usually use really short, often zero-length queues. That way, new threads are spawned immediately when needed.

The problem is that `ScheduledExecutorService` implementations, which OH seems to use "everywhere", can't have a limited queue, because the scheduled tasks are kept in the queue while they wait for their time to execute. As a consequence, queues for `ScheduledExecutorService`s never becomes full, thus they never spawn new threads. I assume that this is the reason why the practice of enabling core threads to time out has arisen. Because, core threads will spawn even if the queue isn't full, but, *they will spawn for every new task submitted, whether they are needed or not*.

If "everything" wasn't a `ScheduledExecutorService`, this wouldn't be an issue. I can't see why the rule thread pool need to be scheduled, for example, I don't think anything is submitted with a schedule there.

For pools where scheduling really is needed, I've made an "experimental solution" that I've only tested slightly, but which I believe could solve the dilemma if it works as intended (it would need a lot of testing before being deployed, but I think the idea is sound):

https://github.com/Nadahar/openhab-core/commit/b44ae751c6ef376fdf6ddb48aa539f68c72caedb

I've called it a "composite executor", because it consists of two. One `ScheduledExecutorService` that only handles the schedules, and one `ExecutorService` that does the actual work. The composite executor "poses" as a `ScheduledExecutorService`, so it can be used like any `ScheduledExecutorService`. The internal `ScheduledExecutorService` only submits tasks to the other executor when they are due, so it should do fine running just a couple of threads, as this is very quick. The internal "work" executor can then be anything, including a dynamic pool that scales as needed.

I think OH's performance and resource use could improve a lot by transitioning from these fixed-sized pools to dynamic ones in many places. It would allow the resources required to scale based on demand, rather than consuming "too much" for an installation that does very little, and "too little" when things get demanding, leading to delays and general "sluggishness".

**Comment by seime** [comment]
My 2 cents:

If we could extend this proposed CompositeExecutorService with 
* the logging added in bundles/org.openhab.core/src/main/java/org/openhab/core/internal/common/WrappedScheduledExecutorService.java
* task "labelling" and expected maximum runtime as overloaded methods like here https://github.com/seime/openhab-esphome/blob/269fe39ba457cee9fa19a4d369945e7e61cf5a0e/src/main/java/no/seime/openhab/binding/esphome/internal/handler/MonitoredScheduledThreadPoolExecutor.java#L132

.. we would get at least these major benefits:

* no more blocked threads causing delayed execution of rules/item updates/etc
* lower memory consumption by not spinning up thousands of threads due to blocked or slow rules blindly (users doing sleep() in rules) - thread stack size memory requests isn't negliglbe 
* logging with stacktrace of tasks that blocks more than a predefined limit - much easier to identify root causes of time-consuming "why is my device reacting after 5-20 seconds instead of immediately" cries for help on the forum
* with task labelling and custom timeouts pr task; 
  * possibility for devs to add expected task runtime (opposed to a fixed 5s limit as of now) for tasks that _do_ by nature take more time (ie network scans)
  * possibility for devs to label such tasks for easier debugging (opposed to decoding stack traces)

IMHO the new executor _should_ have a configurable max threads property with a sensible default, but also log a warning if this limit is hit. This to avoid slowdown due to excessive threads and corresponding memory allocations.

**Comment by kaikreuzer** [comment]
@Nadahar Thanks for bringing this very old topic to the surface again! Your explanations sound logical and it seems that the `ScheduledExecutorService` is then indeed not capable of the desired behavior.
You're suggestion sounds like a good way around that, while again bringing a bit more complexity to it - but that should be fine, if it works as expected. I wonder whether we actually really NEED scheduled tasks "everywhere" or if we should take a step back and check, which of our thread pools actually require scheduling. We could also offer simple non-scheduled pools and others with scheduling. Afair, I once tried that in the past, but then went for "scheduling everywhere", because it wasn't that easy to draw a clear line...

FWIW, in JUPnP we have a fork of the [QueueingThreadPoolExecutor](https://github.com/jupnp/jupnp/blob/main/bundles/org.jupnp/src/main/java/org/jupnp/QueueingThreadPoolExecutor.java) - it might hence also be worth at improving that one as it is heavily used in openHAB.

**Comment by holgerfriedrich** [comment]
My comment might be a bit off topic in this discussion.

Maybe it would be helpful to understand where it helps to use virtual threads instead of thread pools.
Unfortunately, I don't have time to look into it. I already gave it a try for the KNX addon, there should be a branch on my repo.
